### DISCUSSION TITLE
IRQ callback from button press on Lilygo T5-Plus

### DISCUSSION THREAD

**Discussion by brianwyld** [micropython/micropython#12215] [ESP32] [answered]
Hi,

I would like to have a callback when pressing one of the 2 input buttons on a Lilygo T5-Plus board. I can run the examples from the micropython GPIO documentation, everything seems fine, I can read the button values when pressed/released, but my 'handler' callback is never called!!?
Example with STR_IO0 button:
```
>>> import machine
import machine
>>> p = machine.Pin(0,  mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
p = machine.Pin(0,  mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
>>> def callback(p):
def callback(p):
...     print(p)
>>> p.irq(trigger=machine.Pin.IRQ_RISING | machine.Pin.IRQ_FALLING, handler=callback)
p.irq(trigger=machine.Pin.IRQ_RISING | machine.Pin.IRQ_FALLING, handler=callback)
<IRQ>
>>>
>>> p.value()
p.value()
1 
>>> p.value()
p.value()
0 
```
In between the first call to p.value() and the second I am pressing the button, p.value() sees this just fine, but the callback() function is apparently never called??

I feel very stupid.... anyone got an idea of why no callback?

thanks

**Comment by peterhinch** [comment]
Your code runs on a Pyboard (just changing the pin no), with the callback running when I pull the pin down:
```python
>>> import machine
>>> p = machine.Pin('X1',  mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
>>> def callback(p):
...     print(p)
... 
>>> p.irq(trigger=machine.Pin.IRQ_RISING | machine.Pin.IRQ_FALLING, handler=callback)
>>> Pin(Pin.cpu.A0, mode=Pin.IN, pull=Pin.PULL_UP)
Pin(Pin.cpu.A0, mode=Pin.IN, pull=Pin.PULL_UP)
Pin(Pin.cpu.A0, mode=Pin.IN, pull=Pin.PULL_UP)
```
I can only think there's an issue with the board or the IDE.

**Reply by brianwyld** [reply]
OK thanks Peter. At least I haven't completely lost the ability to code...

I'll see if anyone on the lilygo forum has had the problem...(this board has a few other hw issues with the epaper it seems)

**Comment by brianwyld** [comment]
Hi Peter, is the pyboard a ESP32 or an ESP32S3?

The code works fine on an original T5 (ESP32), but not on the T5-Plus (ESP32S3) : tried with IO0, IO21 and IO15 pins....

Maybe there is a fix needed for the S3?

**Reply by andrewleech** [reply]
Just FYI the pyboard uses an stm chip, not an esp.

**Comment by brianwyld** [comment]
Appears to be a bug in ESP IDF 4.4:
https://github.com/micropython/micropython/issues/8509

**Comment by robert-hh** [comment]
The current nightly builds use esp-idf v5.0.2. You can try these,

**Reply by brianwyld** [reply]
thanks, but i build a local version with additional frozen stuff which pulls 1.19.1 to build the bin, so my esp-idf is local.

**Reply by andrewleech** [reply]
In that case hopefully just updating your local micropython repo and idf versions will resolve the issue?

**Reply by brianwyld** [reply]
yes should do. My experience with 'just updating' is that it usually takes me at least a day if not 2 or 3 to get it to build again after I update (but maybe thats due to using WSL2 on a W11 PC...)

**Comment by peterhinch** [comment]
This works on an ESP32-S3 with the latest firmware
```python
import machine
p = machine.Pin(8,  mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
def callback(p):
    print(p)
p.irq(trigger=machine.Pin.IRQ_RISING | machine.Pin.IRQ_FALLING, handler=callback)
```

**Reply by brianwyld** [reply, marked as answer]
thanks Peter. It was indeed a ESP32-S3 IDF 4.4 issue.

**Comment by brianwyld** [comment]
My IRQ callback now works as expected. However, I cannot get it to wake from a lightsleep...

I do;
```
import machine
from machine import Pin
import esp32
p21=Pin(21,mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
p21.irq(handler=print, trigger=(machine.Pin.IRQ_FALLING | machine.Pin.IRQ_RISING), wake=machine.SLEEP)
```
I press the button on IO21, I get the IRQ and it prints 'Pin(21)'. So far so good
```
esp32.wake_on_ext1(pins=[p21], level=esp32.WAKEUP_ALL_LOW)
machine.lightsleep(10000)
```
ESP32S3 sleeps (@6mA) for 10s. Pressing the button has no effect. When it returns from the sleep after the timeout of 10s, it prints 'Pin(21)'.

I see other posts where people have this working, but only with 'normally low' (PULL_DOWN) ios and using WAKE_ANY_HIGH for the level param... IO21 should be an 'RTC' pin so able to wake from light sleep...

Anyone able to shed light on this?

thanks!

**Comment by brianwyld** [comment]
Testing on a ESP32 shows that the wake_on_ext1() works as expected (with the pain in that 'ALL_LOW' means only detects when all the buttons are pressed simultaneously...)
But the ESP32S3 doesn't work with this mode as far as I can see.
```
import machine
from machine import Pin
import esp32
p21=Pin(21,mode=machine.Pin.IN, pull=machine.Pin.PULL_UP)
p21.irq(handler=print, trigger=(machine.Pin.IRQ_FALLING | machine.Pin.IRQ_RISING))
esp32.wake_on_ext1(pins=[p21], level=esp32.WAKEUP_ALL_LOW)
machine.lightsleep(10000)
```
pressing the button IO21 during the 30s of sleep interrupts and calls the callback on an ESP32, but has no effect on ESP32S3.

So, decided to try with the 'gpio wakeup' feature (which is ok as I am using lightsleep() not deepsleep).
Added a esp32ext module, with a wake_on_gpios()  method to call the gpio_wakeup_enable() and esp_sleep_enable_gpio_wakeup() functions. 

This works to exit the lightsleep() when any of the gpio list are pressed!
But now the irq callback for the button is not executed from my app... but is when doing it at the REPL... 
Any pointers on where in the MP code the IRQs are checked post sleep to call the Pin irq? 

PS >(I also implemented a method fo call esp_sleep_get_gpio_wakeup_status(), but for some reason this function is not available (conditionned by requiring that GPIO can wake from DEEP sleep, even though its for light sleep use...)
